### PR TITLE
Adding the new arg parser and hparam format

### PR DESCRIPTION
**Pull request by eelcovdw**
Switching to a new argparser which accepts nested args, which makes the config files and hparams in the codebase easier to use.

- the Hyperparameters class now uses jsonargparse instead of argparse, all functionality to read/write config files is now done by this library.

- config files are now YAML by default.

- the contents of hparams.py are now split over hparams.py for the parser, args.py for all args, args groups and types, and utils.py for other functionality. utils contains a function to convert old json args to the new format.

- All arg changes are stored in hparam_translation_dict.yaml

- Added new demo yaml configs. Not all .sh scripts in demo work, they need manual changes if they contain updated args outside of the config file.

- Changes outside of the hparams folder are to apply the new hparams (ie, hparams.lr_scheduler -> hparams.lr.scheduler)

- Tested with the flickr demo configs for train/translate on aevnmt and conditional models.

### CODE REVIEW

**Comment by Roxot** [comment]
This looks great! Thanks for also providing a conversion script. Please feel free to merge this into master.
